### DISCUSSION TITLE
Unsupported cipher or incorrect key length

### DISCUSSION THREAD

**Discussion by steakhutzeee** [firefly-iii/firefly-iii#6407] [Firefly III]
Hi!

Noted i could not access Firefly III so restarted my container and got the following logs:

[https://gist.github.com/steakhutzeee/36b0d947e86d1430d61be3be1f49a8ec](https://gist.github.com/steakhutzeee/36b0d947e86d1430d61be3be1f49a8ec)

I changed nothing in the Firefly configuration i always use. The only thing i updated my db container that's the linuxserver/mariadb.

Maybe these changes are the cause? https://info.linuxserver.io/issues/2022-08-20-nginx-base/

Thank you in advance!

**Comment by JC5** [comment]
It's usually because the APP_KEY value is not 32 characters long. The Firefly III apps aren't based on nginx.

**Reply by steakhutzeee** [reply]
> It's usually because the APP_KEY value is not 32 characters long. The Firefly III apps aren't based on nginx.

That's strange, i changed nothing.

In my compose i have `env_file: .env` and in that file, the APP_KEY is the same as ever and has 32 char.

What could i do?

**Reply by JC5** [reply]
I'm not really sure. If anything, all commands should fail. Not just some of them. Try deleting the image, make it a fresh start. Be sure to keep your volumes.

This is trial and error, you see.

Are you using Portainer, by any chance?

**Reply by steakhutzeee** [reply]
No I'm using docker compose. Tried with fresh image but issue persists.

**Reply by timaschew** [reply]
Then please provide all commands (download/installation and run) to allow us to verify the issue.

**Reply by steakhutzeee** [reply]
Sure,

here are both my compose and .env file:

https://gist.github.com/steakhutzeee/39b2bf87e8325539beb5069306557e37

https://gist.github.com/steakhutzeee/07071f9516a5f890ad33d534af1ebdf5

The strange point is that i'm running it from a long time and it started complaining about the ciphere just now.

**Reply by JC5** [reply]
I honestly have no clue. Final thing you could try is stop Firefly III, drop the image, prune your images, and redownload.

**Reply by steakhutzeee** [reply]
This is what i did before. Stopping Firefly, docker compose down, remove the image with docker image rm, and docker compose up -d again, so that it pulled it again.

Should do something different?

Maybe replacing the APP_KEY? It's strange it's doing this on it's own.

**Reply by JC5** [reply]
Do you get a different error when you intentionally make it too small or too large?

**Reply by timaschew** [reply]
@steakhutzeee can you please try the default env?

```sh
curl https://raw.githubusercontent.com/firefly-iii/firefly-iii/main/.env.example > .env # keep the file as it is!
curl https://raw.githubusercontent.com/firefly-iii/docker/main/docker-compose.yml > docker-compose.yml
# change port from 80 to anything else if you can't/don't want to use port 80 in the docker-compose.yml file
docker-compose down -v # warning, volumes defined in the compose file are removed, double check if you don't use the volume names in a different compose file as well and need to keep them. In that case just rename the volumes everywhere in the compose file
docker-compose up

```

**Reply by steakhutzeee** [reply]
> Do you get a different error when you intentionally make it too small or too large?

Looks the same, also tried generating a new one.

**Reply by steakhutzeee** [reply]
> try the default env

I tried rebasing my .env on the .env.example but the issue persists. 

You mean creating a new blank container?

**Reply by JC5** [reply]
Sorry if the .env.example doesn't even work, I'm really out of ideas. 

Are you using Portainer by any chance?

**Reply by steakhutzeee** [reply]
No I run it from a docker-compose.yml and run it from command line with docker compose up -d.

I could try with a blank .env maybe? Without editing it.

**Reply by JC5** [reply]
Well if the container doesn't work with the default files something is wrong, but I fear it may be on your side. I just have no idea what it could be.

Try the blank .env, I'm curious!

**Reply by steakhutzeee** [reply]
> Well if the container doesn't work with the default files something is wrong, but I fear it may be on your side. I just have no idea what it could be.
> 
> Try the blank .env, I'm curious!

Using the default .env file i get a lot of errors because it can't see the db but no errors about the cipher.

**Reply by timaschew** [reply]
1. Please show the errors.
2. Did you modify the .env file or docker-compose.yml? If yes, what (which lines) did you change?
3. Did you try out my commands:

```sh
curl https://raw.githubusercontent.com/firefly-iii/firefly-iii/main/.env.example > .env # keep the file as it is!
curl https://raw.githubusercontent.com/firefly-iii/docker/main/docker-compose.yml > docker-compose.yml
# change port from 80 to anything else if you can't/don't want to use port 80 in the docker-compose.yml file
docker-compose down -v # warning, volumes defined in the compose file are removed, double check if you don't use the volume names in a different compose file as well and need to keep them. In that case just rename the volumes everywhere in the compose file
docker-compose up
```

**Reply by steakhutzeee** [reply]
> 1. Please show the errors.
> 
>     2. Did you modify the .env file or docker-compose.yml? If yes, what (which lines) did you change?
> 
>     3. Did you try out my commands:
> 
> 
> ```shell
> curl https://raw.githubusercontent.com/firefly-iii/firefly-iii/main/.env.example > .env # keep the file as it is!
> curl https://raw.githubusercontent.com/firefly-iii/docker/main/docker-compose.yml > docker-compose.yml
> # change port from 80 to anything else if you can't/don't want to use port 80 in the docker-compose.yml file
> docker-compose down -v # warning, volumes defined in the compose file are removed, double check if you don't use the volume names in a different compose file as well and need to keep them. In that case just rename the volumes everywhere in the compose file
> docker-compose up
> ```

No i only tried with the default .env file from the link you provided and the logs only complained about the database not present. 

I also used my APP_KEY in the .env file provided by your link (the file had more data in it, i probably was using an old one). But the issue about the cipher persisted.

What you mean is to run a new test FireflyIII/DB container next to mine?

I already posted my compose, that is https://gist.github.com/steakhutzeee/39b2bf87e8325539beb5069306557e37 and see no structure differences with the default linked.

**Comment by steakhutzeee** [comment]
Ok, this is really strange. I commented the APP_KEY line from the .env file and added it to the docker-compose.yml file directly as an env variable. No issue, the container started flawlessy!

Noted this part at the end of the logs but do not understand why it's there twice, and the client secrets are different:

```
today at 00:42:42  Client ID: 1271
today at 00:42:42  Client secret: xxx
today at 00:42:42  Password grant client created successfully.
today at 00:42:42  Client ID: 1272
today at 00:42:42  Client secret: xxx
today at 00:42:43  Updated version.
```

**Reply by JC5** [reply]
Are you using portainer, by any chance? The two client ID's are indeed there twice, it's not an issue.

**Reply by steakhutzeee** [reply]
No, using docker compose up -d

**Reply by JC5** [reply]
Very strange, :(

**Comment by skorvek** [comment]
I encountered the same error today when updating my firefly container image.  I use portainer with a docker compose file and env file in my private git repository. I changed nothing - just re-pulled the latest image when the error occurred.

What I discovered is that the app key was truncated to only a few characters (due to special characters - '#' in my case which in retrospect commented the rest of the line) which broke laravel.  When I quoted the APP_KEY everything was back to normal. I obviosuly shouldn't have used that character in my faceroll of the keyboard a while back, but it worked just fine then. Something must have changed in docker compose's YAML parser with a recent update.

Perhaps quote the app key in the sample docker-compose file to prevent others from encountering this?

**Reply by JC5** [reply]
The app key isn't listed in the docker compose file, but I can add some instructions to the default `.env` file.

**Reply by steakhutzeee** [reply]
Thanks! But unfortunately my key has no "#" in it.

**Comment by skorvek** [comment]
Yes sorry- the env file referenced from the compose file.
